File: dfs.py
class Node:
    def __init__(self, index):
        self.index = index
        self.adjacency_list = []
        self.visited = False
        self.start_time = None
        self.end_time = None

    def add_edge(self, node):
        self.adjacency_list.append(node)


def dfs(node):
    global time
    node.visited = True
    node.start_time = time
    time += 1

    print("Visiting node:",node.index)
    
    for neighbor in node.adjacency_list:
        if not neighbor.visited:
            dfs(neighbor)

    node.end_time = time
    time += 1


nodes = [Node(i) for i in range(7)]
transposed_nodes= [Node(i) for i in range(7)]

nodes[0].add_edge(nodes[2])
nodes[1].add_edge(nodes[2])
nodes[2].add_edge(nodes[3])
nodes[2].add_edge(nodes[5])
nodes[3].add_edge(nodes[4])
nodes[5].add_edge(nodes[6])

transposed_nodes[2].add_edge(transposed_nodes[0])
transposed_nodes[2].add_edge(transposed_nodes[1])
transposed_nodes[3].add_edge(transposed_nodes[2])
transposed_nodes[5].add_edge(transposed_nodes[2])
transposed_nodes[4].add_edge(transposed_nodes[3])
transposed_nodes[6].add_edge(transposed_nodes[5])

# for node in nodes:
#     node.visited = False

# time = 1
# dfs(nodes[0])

# unvisited_node = None
# for node in nodes:
#     if not node.visited:
#         unvisited_node = node
#         break

# if unvisited_node:
#     dfs(unvisited_node)

# node_timing={}
# for node in nodes:
#     print(node.index,": Start time =",node.start_time,"End time =",node.end_time)
#     node_timing[node.index]=node.end_time

# print(sorted(node_timing, key=lambda k: node_timing[k])[::-1])


for node in transposed_nodes:
    node.visited = False

time = 1
dfs(transposed_nodes[6])

unvisited_node = None
for node in transposed_nodes:
    print("hi")
    if not node.visited:
        unvisited_node = node
        break

if unvisited_node:
    dfs(unvisited_node)

node_timing={}
for node in transposed_nodes:
    print(node.index,": Start time =",node.start_time,"End time =",node.end_time)
    node_timing[node.index]=node.end_time

#print(sorted(node_timing, key=lambda k: node_timing[k])[::-1])

File: dfs.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_VERTICES 100

int time = 1;
int visited[MAX_VERTICES] = {0};
int predecessor[MAX_VERTICES] = {0};
int start_time[MAX_VERTICES] = {0};
int end_time[MAX_VERTICES] = {0};

struct Graph
{
    int numVertices;
    int adjacencyMatrix[MAX_VERTICES][MAX_VERTICES];
};

struct Graph *createGraph(int vertices)
{
    struct Graph *g = (struct Graph *)malloc(sizeof(struct Graph));
    g->numVertices = vertices;
    for (int i = 0; i < g->numVertices; i++)
    {
        for (int j = 0; j < g->numVertices; j++)
        {
            g->adjacencyMatrix[i][j] = 0;
        }
    }
    return g;
}

void addEdge(struct Graph *g, int src, int dest)
{
    g->adjacencyMatrix[src][dest] = 1;
    // For undirected graph
    g->adjacencyMatrix[dest][src] = 1;
}

void dfs(struct Graph *g, int src)
{
    visited[src] = 1;
    predecessor[src] = -1;
    start_time[src] = time;
    time++;
    for (int i = 0; i < g->numVertices; i++)
    {
        if (g->adjacencyMatrix[src][i] == 1 && visited[i] == 0)
        {
            dfs(g, i);
        }
    }
    end_time[src] = time;
    time++;
}

void printfDFS(struct Graph *g)
{
    printf("\nVertex \t visited \t start_time \t end_time");
    for (int i = 0; i < g->numVertices; i++)
    {
        printf("\n%d \t\t %d \t\t %d \t\t %d", i, visited[i], start_time[i], end_time[i]);
    }
}

void printGraph(struct Graph *g)
{
    printf("\nThe Graph is:\n");
    for (int i = 0; i < g->numVertices; i++)
    {
        for (int j = 0; j < g->numVertices; j++)
        {
            printf("%d", g->adjacencyMatrix[i][j]);
        }
        printf("\n");
    }
}

int main()
{
    int v;
    printf("Enter the number of vertices in the graph: ");
    scanf("%d", &v);

    struct Graph *g = createGraph(v);
    addEdge(g, 0, 1);
    addEdge(g, 0, 2);
    addEdge(g, 1, 2);
    addEdge(g, 1, 3);
    addEdge(g, 2, 4);
    addEdge(g, 3, 4);

    printGraph(g);

    dfs(g, 0);

    printfDFS(g);
    return 0;
}